### PR TITLE
feat: implement GraphQL Client (WIP)

### PR DESCRIPTION
**Pull request by chazkiker2**
# Description

- replace manual GraphQL structs with derived `graphql_client::GraphQLQuery`

### CODE REVIEW

**Comment by chazkiker2** [comment]
I was having trouble acquiring the appropriate query data from GitHub while pushing the derived queries from `graphql_client` through to the `octocrab::instance().graphql()` method.

The response contains only errors from GitHub (not data). So the `octo.graphql` method is accepting the `serde_json::to_value(QuerySearch::build_query)` result, but I just cannot figure out how to pass it in in a way that GitHub is okay with. 

### Query Struct

```rust
#[derive(graphql_client::GraphQLQuery)]
#[graphql(
    schema_path = "schema.docs.graphql",
    query_path = "query_1.graphql",
    response_derives = "Serialize,Debug"
)]
struct QuerySearch;
```

### Calling Code
```rust
    let query_string: String = format!(
        r#"repo:{org_name}/{repo_name} is:pr created:>{date_str}"#,
        org_name = org_name,
        repo_name = repo_name,
        date_str = date_str,
    );

    let q = QuerySearch::build_query(query_search::Variables {
        query_string: query_string,
    });
    let q_body = serde_json::to_value(&q)?;
    let q_pretty = serde_json::to_string_pretty(&q_body).unwrap();

    println!("{}", &q_pretty);

    let octo = octocrab::instance();
    let response: Response<query_search::ResponseData> = octo.graphql(&q_pretty).await?;
    println!("{:#?}", response);
```

### This is what `println!("{}", &q_pretty);` prints:

```json
{
  "operationName": "QuerySearch",
  "query": "query QuerySearch($query_string: String!) {\n    search(query:$query_string, type:ISSUE,last:1) {\n        issueCount\n    }\n}\n",
  "variables": {
    "query_string": "repo:rust-lang/rust is:pr created:>2021-05-20"
  }
}
```

### `println!("{}", response);` prints:

```
Response {
    data: None,
    errors: Some(
        [
            Error {
                message: "Parse error on \"operationName\" (STRING) at [2, 3]",
                locations: Some(
                    [
                        Location {
                            line: 2,
                            column: 3,
                        },
                    ],
                ),
                path: None,
                extensions: None,
            },
        ],
    ),
}
```

File: src/metrics/list_repos.rs [hunk 1]
```diff
@@ -95,26 +111,18 @@ async fn count_pull_requests_graphql(
...
     );
-
+    let q = QuerySearch::build_query(query_search::Variables {
+        query_string: query_string,
+    });
     let octo = octocrab::instance();
-    let response: Response<Data> = octo.graphql(&query_string).await?;
-
-    response.data.search.issue_count as usize
+    let response: Response<query_search::ResponseData> = octo.graphql_with_params(&q).await?;
```
Comment (nikomatsakis) [inline src/metrics/list_repos.rs:124]: I was imagining we might make this

```suggestion
    let response: Response<query_search::ResponseData> = octo.graphql_with_params::<QuerySearch>(query_search::Variables { query_String }).await?;
```

and have the extension function do the work of invoking `QuerySearch::build_query`.

File: src/metrics/list_repos.rs [hunk 2]
```diff
@@ -64,15 +65,30 @@ impl Producer for ListReposForOrg {
...
+#[graphql(
+    schema_path = "gql/schema.docs.graphql",
+    query_path = "gql/query_search.graphql",
+    response_derives = "Serialize,Debug"
+)]
+struct QuerySearch;
+
+#[async_trait]
+pub trait GQL {
+    async fn graphql_with_params<R: octocrab::FromResponse + Send>(
```
Comment (nikomatsakis) [inline src/metrics/list_repos.rs:78]: Maybe this?

```suggestion
    async fn graphql_with_params<Q>(variables: Q::Variables) -> octocrab::Result<Q::ResponseData>
    where
        Q::Variables: Send,
    {
        ...
    }
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/metrics/list_repos.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/metrics/list_repos.rs:124]: An alternative might be something like

```rust
QuerySearch::build(...).execute(&octo)
```

In which case, we would implement the "extension trait" for [`QueryBody<Q>`](https://docs.rs/graphql_client/0.9.0/graphql_client/struct.QueryBody.html).

That might be more ergonomic and better?

File: src/metrics.rs [hunk 3]
```diff
@@ -29,118 +28,75 @@ pub use export_to_sheets::ExportToSheets;
...
 
-#[derive(Deserialize, Debug)]
-struct RepoNode {
-    name: String,
+#[async_trait]
+impl<V> GQL for QueryBody<V>
+where
+    V: Serialize + Send + Sync,
+{
+    async fn execute<Q: GraphQLQuery + Send>(self) -> octocrab::Result<Response<Q::ResponseData>> {
```
Comment (nikomatsakis) [inline src/metrics.rs:57]: Hmm, it's annoying that you need the `Q` parameter here! Now I'm thinking we should refactor this once more (sorry), to something like this:

```rust
impl<Q> GQL for Q
where
    Q: GraphQLQuery,
{
    fn execute(variables: Self::Variables) -> octocrab::Result<Response<Self::ResponseData>> {
        let body = Self::build(variables);
        Ok(octocrab::instance().post("graphql", Some(&body)).await?)
    }
}
```

File: src/metrics.rs [hunk 4]
```diff
@@ -29,118 +28,75 @@ pub use export_to_sheets::ExportToSheets;
...
-            repos_data
-                .edges
-                .iter()
-                .map(|edge| edge.node.name.to_owned()),
-        );
+        let res = OrgRepos::build_query(org_repos::Variables {
+            org_name: org_name.to_owned(),
+            after_cursor,
+        })
+        .execute::<OrgRepos>()
```
Comment (nikomatsakis) [inline src/metrics.rs:73]:
```suggestion
        let res = OrgRepos::execute(org_repos::Variables {
            org_name: org_name.to_owned(),
            after_cursor,
        })
```

Comment (nikomatsakis) [reply src/metrics.rs:73]: (this is what we would do if we took my approach above)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/metrics.rs (hunk 3 above)
Comment (chazkiker2) [reply src/metrics.rs:57]: how would you denote this in the trait itself??

Would I need `pub trait GQL<Q>`, essentially? I'm not entirely sure how to move that type parameter that was attached to `execute`

Comment (chazkiker2) [reply src/metrics.rs:57]: jk I think I figured it out

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**